### PR TITLE
feat: Update Anthropic/Cohere for tools use

### PR DESCRIPTION
**Pull request by vblagoje**
### Why:
Adds support for tools in Anthropic and Cohere chat generators

### What:
1. **Anthropic Integration Enhancements:**
   - Introduced handling for `tools` and `tool_choice` parameters in the Anthropic API, enriching the chat responses with the ability to utilize external tools.
   - Added a mechanism to ignore "thinking messages" from tools, improving clarity in chat responses.
   - Expanded unit tests to cover new functionalities, ensuring robustness and reliability.

2. **Cohere Integration Enhancements:**
   - Adjusted chat message building logic to accommodate tool calls
   - Updated unit tests to assess tool use functionality, ensuring the new features work as intended.

3. **General Maintenance:**
   - A test file in the Cohere integration was renamed for consistency, aligning with the naming convention of other test files in the project.

### How can it be used:
- The updates allow developers integrating Anthropic or Cohere-based chat functionalities to:
  - Configure tool use directly within chat generators for enriched responses.
  - Dynamically filter auxiliary "thinking messages" for clearer communication.
  - Seamlessly include tool-generated content within chat interfaces, making it possible to offer advanced, AI-driven interactions, such as fetching live data or performing complex computations on behalf of the user.
- Example usage (pseudocode):
```python
# For Anthropic:
chat_generator = AnthropicChatGenerator(api_key="API_KEY", ignore_tools_thinking_messages=True)
response = chat_generator.run(messages=["User message"], generation_kwargs={"tools": [tools_schema]})

# For Cohere:
chat_generator = CohereChatGenerator(model="MODEL_NAME")
response = chat_generator.run(messages=["User message"], generation_kwargs={"tools": [tools_schema]})
```

### How did you test it:
- The enhancements were tested through unit and integration tests covering:
  - Initialization of chat generator instances with new parameters.
  - Conversion of tool use blocks to chat messages, ensuring the logical flow and data structure are preserved.
  - Interaction scenarios invoking tool use, verifying the correct representation and inclusion of tool-generated content in chat responses.

### Notes for the reviewer:
- Reviews details of tools usage, make sure we didn't forget something

### CODE REVIEW

**Comment by vblagoje** [comment]
@davidsbatista I added @anakin87 for this review as he's familiar with tool calling in these models 🙏

**Review by anakin87 — COMMENTED** [review]
In general, the changes look good.

---

I'm not sure about this snippet:
```python
chat_generator = CohereChatGenerator(model="MODEL_NAME")
response = chat_generator.run(messages=["User message"], generation_kwargs={"tools": [tools_schema]})
```

I think we should introduce some abstraction to provide tools to our generators.
From a semantic point of view, the tools are not `generation_kwargs` like temperature, top_p, etc.

**Comment by vblagoje** [comment]
> In general, the changes look good.
> 
> I'm not sure about this snippet:
> 
> ```python
> chat_generator = CohereChatGenerator(model="MODEL_NAME")
> response = chat_generator.run(messages=["User message"], generation_kwargs={"tools": [tools_schema]})
> ```
> 
> I think we should introduce some abstraction to provide tools to our generators. From a semantic point of view, the tools are not `generation_kwargs` like temperature, top_p, etc.

Do you mean like another dataclass that defines tools @anakin87 ? But this PR is not the right place for such change, we should first understand impact across all generators, figure out how to do this properly, and roll it out it in a seperate PR or PRS, not here in this one.

File: integrations/anthropic/src/haystack_integrations/components/generators/anthropic/chat/chat_generator.py [hunk 1]
```diff
@@ -203,18 +211,24 @@ def run(self, messages: List[ChatMessage], generation_kwargs: Optional[Dict[str,
             completions = [self._connect_chunks(chunks, start_event, delta)]
         # if streaming is disabled, the response is an Anthropic Message
         elif isinstance(response, Message):
+            has_tools_msgs = any(isinstance(content_block, ToolUseBlock) for content_block in response.content)
+            if has_tools_msgs and self.ignore_tools_thinking_messages:
+                response.content = [block for block in response.content if isinstance(block, ToolUseBlock)]
```
[the remark this answers is not in the document]
Comment (vblagoje) [reply integrations/anthropic/src/haystack_integrations/components/generators/anthropic/chat/chat_generator.py:216]: Yes, good question. No support yet for FC under streaming generators. I think we should mention that somewhere and make it across all generators.

File: integrations/anthropic/src/haystack_integrations/components/generators/anthropic/chat/chat_generator.py [hunk 2]
```diff
@@ -95,13 +100,18 @@ def __init__(
             - `temperature`: The temperature to use for sampling.
             - `top_p`: The top_p value to use for nucleus sampling.
             - `top_k`: The top_k value to use for top-k sampling.
-
+        :param ignore_tools_thinking_messages: Anthropic's approach to tools (function calling) resolution involves a
+        "chain of thought" messages before returning the actual function names and parameters in a message. If
+        ignore_tools_thinking_messages is True, the generator will drop so-called thinking messages when tool
+        use is detected. See the Anthropic [tools](https://docs.anthropic.com/en/docs/tool-use#chain-of-thought-tool-use)
+        for more details.
```
Comment (anakin87) [inline integrations/anthropic/src/haystack_integrations/components/generators/anthropic/chat/chat_generator.py:107]: Probably this should be indented.
Let's also put `include ignore_tools_thinking_messages` and `True` inside backticks

**Comment by vblagoje** [comment]
> I only left a comment on docstrings format.
> 
> Let's merge this PR, keeping in mind that:
> 
> * we probably need an abstraction for tools and way to pass them to LLM calls
> * function calling is not supported when streaming is enabled (please create an issue to keep track of this)

Agreed, here is the Feature request https://github.com/deepset-ai/haystack-core-integrations/issues/813

**Pull request merged**